### PR TITLE
Day5 Ex1 RecyclerView

### PR DESCRIPTION
**Pull request by at-hoavo**
em nộp bài tập phần recyclerview. anh xem  giùm em với!

### CODE REVIEW

File: app/src/main/res/drawable/custom_button_friends.xml [hunk 1]
```diff
@@ -0,0 +1,8 @@
+<?xml version="1.0" encoding="utf-8"?>
+<selector xmlns:android="http://schemas.android.com/apk/res/android">
+<item>
+    <shape>
+        <solid android:color="#FF0033"/>
+    </shape>
+</item>
+</selector>
```
Comment (at-tamnguyen2) [inline app/src/main/res/drawable/custom_button_friends.xml]: please insert break line

File: app/src/main/res/layout/fragment_friends.xml [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+    android:layout_width="match_parent"
+    android:layout_height="match_parent"
+    android:orientation="vertical">
+
+    <android.support.v7.widget.RecyclerView
+        android:id="@+id/recyclerView"
+        android:layout_width="match_parent"
+        android:layout_height="match_parent" />
+
+</LinearLayout>
```
Comment (at-tamnguyen2) [inline app/src/main/res/layout/fragment_friends.xml]: please insert break line, and beacause this layout only one view -> remove LinearLayout

File: app/src/main/res/values/dimen.xml [hunk 3]
```diff
@@ -0,0 +1,11 @@
...
+<resources>
+    <dimen name="relativelayout.layout_height">80dp</dimen>
+    <dimen name="imageview.layout_marginStart">5dp</dimen>
+    <dimen name="textview.name.layout_marginStart">75dp</dimen>
+    <dimen name="linearlayout.layout_marginEnd">10dp</dimen>
+    <dimen name="button.add.layout_marginEnd">5dp</dimen>
+    <dimen name="button.friends.layout_height">40dp</dimen>
+    <dimen name="button.friends.layout_margin_Start">210dp</dimen>
+    <dimen name="button.friends.layout_marginTop">5dp</dimen>
+</resources>
```
Comment (at-tamnguyen2) [inline app/src/main/res/values/dimen.xml]: please insert break line

File: app/src/main/java/vn/asiantech/internship/FriendsFragment.java [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+import android.view.LayoutInflater;
+import android.view.View;
+import android.view.ViewGroup;
+
+import java.util.ArrayList;
+import java.util.Collections;
+
+public class FriendsFragment extends Fragment {
+    RecyclerView recyclerView;
+    private ArrayList<String> mListUsers = new ArrayList<>();
```
Comment (at-tamnguyen2) [inline app/src/main/java/vn/asiantech/internship/FriendsFragment.java]: please change mListUsers to mUsers

File: app/src/main/java/vn/asiantech/internship/FriendsFragment.java [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+
+    @Nullable
+    @Override
+    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, Bundle savedInstanceState) {
+        View v = inflater.inflate(R.layout.fragment_friends, container, false);
+        String[] list = getResources().getStringArray(R.array.list_users);
```
Comment (at-tamnguyen2) [inline app/src/main/java/vn/asiantech/internship/FriendsFragment.java]: please change list to users

File: app/src/main/java/vn/asiantech/internship/FriendsFragment.java [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+import android.support.v7.widget.RecyclerView;
+import android.view.LayoutInflater;
+import android.view.View;
+import android.view.ViewGroup;
+
+import java.util.ArrayList;
+import java.util.Collections;
+
+public class FriendsFragment extends Fragment {
+    RecyclerView recyclerView;
```
Comment (at-tamnguyen2) [inline app/src/main/java/vn/asiantech/internship/FriendsFragment.java]: change "recyclerView" to "mRecyclerView"

File: app/src/main/java/vn/asiantech/internship/FriendsFragment.java [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+    @Override
+    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, Bundle savedInstanceState) {
+        View v = inflater.inflate(R.layout.fragment_friends, container, false);
+        String[] list = getResources().getStringArray(R.array.list_users);
+        Collections.addAll(mListUsers, list);
+        recyclerView = (RecyclerView) v.findViewById(R.id.recyclerView);
+        AdapterRecycler adapter = new AdapterRecycler(mListUsers);
+        recyclerView.setAdapter(adapter);
+        LinearLayoutManager linearLayoutManager = new LinearLayoutManager(getActivity());
+        recyclerView.setLayoutManager(linearLayoutManager);
```
Comment (at-tamnguyen2) [inline app/src/main/java/vn/asiantech/internship/FriendsFragment.java]: please move code line 28, 29 to before setAdapter for recyclerView (line 27)

File: app/src/main/res/drawable/custom_button_friends.xml [hunk 8]
```diff
@@ -0,0 +1,8 @@
+<?xml version="1.0" encoding="utf-8"?>
+<selector xmlns:android="http://schemas.android.com/apk/res/android">
+<item>
+    <shape>
+        <solid android:color="#FF0033"/>
+    </shape>
+</item>
```
Comment (at-tamnguyen2) [inline app/src/main/res/drawable/custom_button_friends.xml]: why don't use setBackground for button,  please check code and remove this file

File: app/src/main/java/vn/asiantech/internship/Day5/AdapterRecycler.java [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+import android.view.View;
+import android.view.ViewGroup;
+import android.widget.Button;
+import android.widget.TextView;
+
+import java.util.ArrayList;
+
+import vn.asiantech.internship.R;
+
+class AdapterRecycler extends RecyclerView.Adapter<AdapterRecycler.RecyclerViewHoder> {
```
Comment (at-vinhhuynh) [inline app/src/main/java/vn/asiantech/internship/Day5/AdapterRecycler.java]: add java doc

File: app/src/main/java/vn/asiantech/internship/Day5/AdapterRecycler.java [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+        }
+    }
+
+    @Override
+    public int getItemCount() {
+        return mFriends.size();
+    }
+
+    class RecyclerViewHoder extends RecyclerView.ViewHolder {
+        TextView tvName;
```
Comment (at-vinhhuynh) [inline app/src/main/java/vn/asiantech/internship/Day5/AdapterRecycler.java]: add `private` key for all variable
and fix variable name follow convention

File: app/src/main/java/vn/asiantech/internship/Day5/FriendsActivity.java [hunk 11]
```diff
@@ -1,13 +1,15 @@
-package vn.asiantech.internship;
+package vn.asiantech.internship.Day5;
 
 import android.support.v7.app.AppCompatActivity;
 import android.os.Bundle;
 
-public class MainActivity extends AppCompatActivity {
+import vn.asiantech.internship.R;
+
+public class FriendsActivity extends AppCompatActivity {
```
Comment (at-vinhhuynh) [inline app/src/main/java/vn/asiantech/internship/Day5/FriendsActivity.java]: add java doc

File: app/src/main/java/vn/asiantech/internship/Day5/AdapterRecycler.java [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+            holder.btnFriends.setVisibility(View.INVISIBLE);
+        }
+    }
+
+    @Override
+    public int getItemCount() {
+        return mFriends.size();
+    }
+
+    class RecyclerViewHoder extends RecyclerView.ViewHolder {
```
Comment (at-vinhhuynh) [inline app/src/main/java/vn/asiantech/internship/Day5/AdapterRecycler.java]: add java doc

File: app/src/main/java/vn/asiantech/internship/Day5/FriendsFragment.java [hunk 13]
```diff
@@ -0,0 +1,34 @@
...
+import android.view.LayoutInflater;
+import android.view.View;
+import android.view.ViewGroup;
+
+import java.util.ArrayList;
+import java.util.Collections;
+
+import vn.asiantech.internship.R;
+
+public class FriendsFragment extends Fragment {
```
Comment (at-vinhhuynh) [inline app/src/main/java/vn/asiantech/internship/Day5/FriendsFragment.java]: add java doc

File: app/src/main/java/vn/asiantech/internship/Day5/FriendsFragment.java [hunk 14]
```diff
@@ -0,0 +1,34 @@
...
+    private RecyclerView mRecyclerView;
+    private ArrayList<String> mUsers = new ArrayList<>();
```
Comment (at-vinhhuynh) [inline app/src/main/java/vn/asiantech/internship/Day5/FriendsFragment.java]: please use `List<>` instead for `ArrayList<>`

File: app/src/main/res/layout/item_list_friends.xml [hunk 15]
```diff
@@ -0,0 +1,56 @@
+<?xml version="1.0" encoding="utf-8"?>
+<RelativeLayout xmlns:android="http://schemas.android.com/apk/res/android"
+    android:layout_width="match_parent"
+    android:layout_height="@dimen/relativelayout.layout_height"
+    android:orientation="horizontal">
```
Comment (at-vinhhuynh) [inline app/src/main/res/layout/item_list_friends.xml]: Used `RelativeLayout` please remove `android:orientation="horizontal"`

File: app/src/main/res/layout/item_list_friends.xml [hunk 16]
```diff
@@ -0,0 +1,56 @@
...
+            android:gravity="end"
+            android:orientation="vertical">
+
+            <Button
+                android:id="@+id/btnAdd"
+                android:layout_width="wrap_content"
+                android:layout_height="wrap_content"
+                android:layout_marginEnd="@dimen/button.add.layout_marginEnd"
+                android:text="@string/button.text.add"
+                android:visibility="visible" />
```
Comment (at-vinhhuynh) [inline app/src/main/res/layout/item_list_friends.xml]: remove `android:visibility="visible"`

File: app/src/main/java/vn/asiantech/internship/Day5/FriendsFragment.java (hunk 14 above)
Comment (at-hoavo) [inline app/src/main/java/vn/asiantech/internship/Day5/FriendsFragment.java]: can I ask why ArrayList<> doesn't allow?

File: app/src/main/java/vn/asiantech/internship/Day5/AdapterRecycler.java [hunk 17]
```diff
@@ -0,0 +1,85 @@
...
+    }
+
+    @Override
+    public RecyclerViewHoder onCreateViewHolder(ViewGroup parent, int viewType) {
+        LayoutInflater inflater = LayoutInflater.from(parent.getContext());
+        return new RecyclerViewHoder(inflater.inflate(R.layout.item_list_friends, parent, false));
+    }
+
+    @Override
+    public long getItemId(int position) {
```
Comment (at-vinhhuynh) [inline app/src/main/java/vn/asiantech/internship/Day5/AdapterRecycler.java]: Please remove this method if you don't need

File: app/src/main/java/vn/asiantech/internship/Day5/AdapterRecycler.java [hunk 18]
```diff
@@ -0,0 +1,85 @@
...
+
+    @Override
+    public int getItemCount() {
+        return mFriends.size();
+    }
+
+    /**
+     * Use to create a RecyclerViewHoder
+     */
+    class RecyclerViewHoder extends RecyclerView.ViewHolder {
```
Comment (at-vinhhuynh) [inline app/src/main/java/vn/asiantech/internship/Day5/AdapterRecycler.java]: Change class name to `RecyclerViewHolder`

File: app/src/main/java/vn/asiantech/internship/Day5/FriendsActivity.java [hunk 19]
```diff
@@ -0,0 +1,22 @@
...
+import vn.asiantech.internship.R;
+
+/**
+ * Activity contain Fragment to display RecyclerView
+ *
+ * @author at-HoaVo
+ * @version 1.0
+ * @since 2017-6-12
+ */
+public class FriendsActivity extends AppCompatActivity {
```
Comment (at-vinhhuynh) [inline app/src/main/java/vn/asiantech/internship/Day5/FriendsActivity.java]: change class name to `FriendActivity`

File: app/src/main/java/vn/asiantech/internship/Day5/FriendsFragment.java [hunk 20]
```diff
@@ -0,0 +1,41 @@
...
+import vn.asiantech.internship.R;
+
+/**
+ * Fragment contain RecyclerView
+ *
+ * @author at-HoaVo
+ * @version 1.0
+ * @since 2017-6-12
+ */
+public class FriendsFragment extends Fragment {
```
Comment (at-vinhhuynh) [inline app/src/main/java/vn/asiantech/internship/Day5/FriendsFragment.java]: change class name to `FriendFragment`

File: app/src/main/java/vn/asiantech/internship/Day5/FriendsFragment.java [hunk 21]
```diff
@@ -0,0 +1,41 @@
...
+    private RecyclerView mRecyclerView;
+    private ArrayList<String> mUsers = new ArrayList<>();
```
Comment (at-vinhhuynh) [inline app/src/main/java/vn/asiantech/internship/Day5/FriendsFragment.java]: use `List<>` instead for `ArrayList<>`

File: app/src/main/java/vn/asiantech/internship/Day5/FriendsFragment.java [hunk 22]
```diff
@@ -0,0 +1,41 @@
...
+
+    @Nullable
+    @Override
+    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, Bundle savedInstanceState) {
+        View v = inflater.inflate(R.layout.fragment_friends, container, false);
+        String[] users = getResources().getStringArray(R.array.list_users);
+        Collections.addAll(mUsers, users);
```
Comment (at-vinhhuynh) [inline app/src/main/java/vn/asiantech/internship/Day5/FriendsFragment.java]: Use `mUsers.addAll(users);`

File: app/src/main/res/layout/activity_friends.xml [hunk 23]
```diff
@@ -0,0 +1,14 @@
+<?xml version="1.0" encoding="utf-8"?>
+<LinearLayout xmlns:android="http://schemas.android.com/apk/res/android"
+    xmlns:tools="http://schemas.android.com/tools"
+    android:layout_width="match_parent"
+    android:layout_height="match_parent"
+    tools:context="vn.asiantech.internship.Day5.FriendsActivity">
+
+    <fragment
+        android:id="@+id/fragmentFriends"
```
Comment (at-vinhhuynh) [inline app/src/main/res/layout/activity_friends.xml]: change id to `fragmentFriend`

File: app/src/main/res/layout/activity_friends.xml [hunk 24]
```diff
@@ -0,0 +1,14 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Comment (at-vinhhuynh) [inline app/src/main/res/layout/activity_friends.xml]: change file name to `activity_friend`

File: app/src/main/res/layout/fragment_friends.xml [hunk 25]
```diff
@@ -0,0 +1,5 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Comment (at-vinhhuynh) [inline app/src/main/res/layout/fragment_friends.xml]: change file name to `fragment_friend`

File: app/src/main/res/layout/item_list_friends.xml [hunk 26]
```diff
@@ -0,0 +1,54 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Comment (at-vinhhuynh) [inline app/src/main/res/layout/item_list_friends.xml]: change file name to `item_list_friend`

File: app/src/main/res/layout/item_list_friends.xml [hunk 27]
```diff
@@ -0,0 +1,54 @@
...
+<RelativeLayout xmlns:android="http://schemas.android.com/apk/res/android"
+    android:layout_width="match_parent"
+    android:layout_height="@dimen/relativelayout.layout_height">
+
+    <ImageView
+        android:id="@+id/imgView"
```
Comment (at-vinhhuynh) [inline app/src/main/res/layout/item_list_friends.xml]: change id to `imgUser`

File: app/src/main/res/layout/item_list_friends.xml [hunk 28]
```diff
@@ -0,0 +1,54 @@
...
+                android:layout_width="wrap_content"
+                android:layout_height="wrap_content"
+                android:layout_marginEnd="@dimen/button.add.layout_marginEnd"
+                android:text="@string/button.text.add" />
+        </LinearLayout>
+
+    </LinearLayout>
+
+    <Button
+        android:id="@+id/btnFriends"
```
Comment (at-vinhhuynh) [inline app/src/main/res/layout/item_list_friends.xml]: change id to `btnFriend`

File: app/src/main/res/values/strings.xml [hunk 29]
```diff
@@ -1,3 +1,30 @@
 <resources>
-    <string name="app_name">Internship</string>
+    <string name="app_name">Friends</string>
+    <string name="imageview.description">this is image user</string>
+    <string name="button.text.add">+ ADD</string>
+    <string name="button.text.friends">Friends</string>
+    <string-array name="list_users">
```
Comment (at-vinhhuynh) [inline app/src/main/res/values/strings.xml]: change array name to `users` or `list_user`

**Comment by at-hoavo** [comment]
can I ask why in file  FriendFragment.java,  ArrayList<> doesn't allow?

File: app/src/main/java/vn/asiantech/internship/Day5/AdapterRecycler.java [hunk 30]
```diff
@@ -0,0 +1,80 @@
...
+        return mFriends.size();
+    }
+
+    /**
+     * Use to create a RecyclerViewHoder
+     */
+    class RecyclerViewHolder extends RecyclerView.ViewHolder {
+        private TextView mTvName;
+        private Button mBtnAdd;
+        Button mBtnFriends;
```
Comment (at-vinhhuynh) [inline app/src/main/java/vn/asiantech/internship/Day5/AdapterRecycler.java]: add `private`

File: app/src/main/java/vn/asiantech/internship/Day5/AdapterRecycler.java [hunk 31]
```diff
@@ -0,0 +1,80 @@
...
+import vn.asiantech.internship.R;
+
+/**
+ * Use to create a RecyclerView
+ *
+ * @author at-HoaVo
+ * @version 1.0
+ * @since 2017-6-12
+ */
+class AdapterRecycler extends RecyclerView.Adapter<AdapterRecycler.RecyclerViewHolder> {
```
Comment (at-vinhhuynh) [inline app/src/main/java/vn/asiantech/internship/Day5/AdapterRecycler.java]: Class name to `RecyclerViewAdapter`

File: app/src/main/java/vn/asiantech/internship/Day5/AdapterRecycler.java [hunk 32]
```diff
@@ -0,0 +1,80 @@
...
+        return mFriends.size();
+    }
+
+    /**
+     * Use to create a RecyclerViewHoder
+     */
+    class RecyclerViewHolder extends RecyclerView.ViewHolder {
+        private TextView mTvName;
+        private Button mBtnAdd;
+        private Button mBtnFriends;
```
Comment (at-vinhhuynh) [inline app/src/main/java/vn/asiantech/internship/Day5/AdapterRecycler.java]: variable name `mBtnFriend`

File: app/src/main/java/vn/asiantech/internship/Day5/FriendFragment.java [hunk 33]
```diff
@@ -0,0 +1,42 @@
...
+public class FriendFragment extends Fragment {
+    private RecyclerView mRecyclerView;
+    private List<String> mUsers = new ArrayList<>();
+
+    @Nullable
+    @Override
+    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, Bundle savedInstanceState) {
+        View v = inflater.inflate(R.layout.fragment_friend, container, false);
+        String[] users = getResources().getStringArray(R.array.list_user);
+        for (int i = 0; i < users.length; i++)
```
Comment (at-vinhhuynh) [inline app/src/main/java/vn/asiantech/internship/Day5/FriendFragment.java]: Remove this `for-loop`, use `Collections.addAll(mUsers, users);`

**Comment by at-vinhhuynh** [comment]
@at-hoavo 
Thanks for fixing!
LGTM!

**Pull request merged**